File: apps/windows/Taskly/Services/DateParser.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Taskly.Services;

/// <summary>
/// Natural-language date/time parser (CLI-SPEC.md §--due). Port of the
/// proven 0.6.x implementation — grammar is a cross-platform contract:
///   +10m, +2h, +1d, +1w, +1M (30 days; case-sensitive m/M, digits optional)
///   @now, @10am, @10:30am, @22:30, optional tomorrow/tmw/mon..sun modifier
///   absolute: yyyy-MM-dd / yyyy/MM/dd / MM/dd/yyyy / dd/MM/yyyy (1900-2100)
/// Results: 'yyyy-MM-dd HH:mm:ss' or date-only 'yyyy-MM-dd'.
/// </summary>
public sealed class DateParser
{
    private static readonly Regex RelativeDateRegex = new(
        @"^(\d*)([mhdwM])$",
        RegexOptions.CultureInvariant);

    private static readonly Regex TimeRegex = new(
        @"^(\d{1,2})(?::(\d{2}))?(am|pm)?$",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

    private static readonly Regex ExtractRelativeRegex = new(
        @"(?:^|\s)(\+\d+[mhdwM])(?:\s|$)",
        RegexOptions.CultureInvariant);

    private static readonly Regex ExtractAtRegex = new(
        @"@(?:now|\d{1,2}(?::\d{2})?(?:am|pm)?)(?:\s+(?:tomorrow|tmw|mon|tue|wed|thu|fri|sat|sun))?$",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

    public string? Parse(string input)
    {
        if (string.IsNullOrWhiteSpace(input))
        {
            return null;
        }

        var s = input.Trim();

        if (s.StartsWith('+'))
        {
            return ParseRelativeDate(s);
        }

        if (s.StartsWith('@'))
        {
            return ParseAtTime(s[1..]);
        }

        return ParseAbsoluteDate(s);
    }

    /// <summary>Pulls a trailing time command out of text.</summary>
    public (string Text, string? TimeCommand) ExtractTimeCommand(string input)
    {
        if (string.IsNullOrWhiteSpace(input))
        {
            return (input ?? string.Empty, null);
        }

        var text = input;
        string? command = null;

        var atMatch = ExtractAtRegex.Match(text);
        if (atMatch.Success)
        {
            var atIndex = text.IndexOf('@');
            command = text[atIndex..].Trim();
            text = text[..atIndex].TrimEnd();
        }

        var relMatch = ExtractRelativeRegex.Match(text);
        if (relMatch.Success)
        {
            command = relMatch.Groups[1].Value;
            text = text.Remove(relMatch.Index, relMatch.Length).Trim();
        }

        return (text.Trim(), command);
    }

    private string? ParseRelativeDate(string input)
    {
        var body = input[1..];
        var match = RelativeDateRegex.Match(body);
        if (!match.Success)
        {
            return null;
        }

        var num = match.Groups[1].Value;
        var amount = string.IsNullOrEmpty(num) ? 1 : int.Parse(num, CultureInfo.InvariantCulture);
        var unit = match.Groups[2].Value;

        var now = DateTime.Now;
        var result = unit switch
        {
            "m" => now.AddMinutes(amount),
            "h" => now.AddHours(amount),
            "d" => now.AddDays(amount),
            "w" => now.AddDays(amount * 7),
            "M" => now.AddDays(amount * 30),
            _ => (DateTime?)null,
        };

        return result?.ToString("yyyy-MM-dd HH:mm:ss", CultureInfo.InvariantCulture);
    }

    private string? ParseAtTime(string input)
    {
        var trimmed = input.Trim();

        if (trimmed.Equals("now", StringComparison.OrdinalIgnoreCase))
        {
            return DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss", CultureInfo.InvariantCulture);
        }

        var parts = trimmed.Split(new[] { ' ' }, 2, StringSplitOptions.RemoveEmptyEntries);
        var timePart = parts[0];
        var modifier = parts.Length > 1 ? parts[1] : null;

        var match = TimeRegex.Match(timePart);
        if (!match.Success)
        {
            return null;
        }

        var hour = int.Parse(match.Groups[1].Value, CultureInfo.InvariantCulture);
        var min = match.Groups[2].Success ? int.Parse(match.Groups[2].Value, CultureInfo.InvariantCulture) : 0;
        var ampm = match.Groups[3].Success ? match.Groups[3].Value.ToLowerInvariant() : null;

        if (ampm == "am")
        {
            if (hour == 12)
            {
                hour = 0;
            }
        }
        else if (ampm == "pm")
        {
            if (hour < 12)
            {
                hour += 12;
            }
        }

        if (hour is < 0 or > 23 || min is < 0 or > 59)
        {
            return null;
        }

        var today = DateTime.Now;
        var date = new DateTime(today.Year, today.Month, today.Day, hour, min, 0);

        if (!string.IsNullOrEmpty(modifier))
        {
            date = ApplyDayModifier(date, modifier!);
        }
        else if (date < DateTime.Now)
        {
            date = date.AddDays(1);
        }

        return date.ToString("yyyy-MM-dd HH:mm:ss", CultureInfo.InvariantCulture);
    }

    private DateTime ApplyDayModifier(DateTime date, string modifier)
    {
        var m = modifier.ToLowerInvariant();
        if (m is "tomorrow" or "tmw")
        {
            return date.AddDays(1);
        }

        var targetDay = m switch
        {
            "sun" => DayOfWeek.Sunday,
            "mon" => DayOfWeek.Monday,
            "tue" => DayOfWeek.Tuesday,
            "wed" => DayOfWeek.Wednesday,
            "thu" => DayOfWeek.Thursday,
            "fri" => DayOfWeek.Friday,
            "sat" => DayOfWeek.Saturday,
            _ => (DayOfWeek?)null,
        };

        if (targetDay is { } td)
        {
            var diff = ((int)td - (int)date.DayOfWeek + 7) % 7;
            if (diff <= 0)
            {
                diff += 7;
            }

            return date.AddDays(diff);
        }

        return date;
    }

    private string? ParseAbsoluteDate(string input)
    {
        var s = input.Trim();

        string[]? formats =
        {
            "yyyy-MM-dd", "yyyy/MM/dd", "MM/dd/yyyy", "dd/MM/yyyy",
        };

        foreach (var fmt in formats)
        {
            if (DateTime.TryParseExact(s, fmt, CultureInfo.InvariantCulture, DateTimeStyles.None, out var date))
            {
                if (date.Year < ValidationHelper.MinYear || date.Year > ValidationHelper.MaxYear)
                {
                    return null;
                }

                return date.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
            }
        }

        return null;
    }

    // ---------------- display helpers ----------------

    public static string? ExtractDateOnly(string? dateTimeStr)
    {
        if (string.IsNullOrEmpty(dateTimeStr))
        {
            return null;
        }

        return dateTimeStr!.Length >= 10 ? dateTimeStr[..10] : dateTimeStr;
    }

    public static string? ExtractTimeOnly(string? dateTimeStr)
    {
        if (string.IsNullOrEmpty(dateTimeStr) || dateTimeStr!.Length < 16)
        {
            return null;
        }

        return dateTimeStr[11..16];
    }

    public static string CombineDateTime(string? dateStr, string? timeStr)
    {
        var date = ExtractDateOnly(dateStr) ?? DateTime.Now.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
        var time = !string.IsNullOrEmpty(timeStr) ? timeStr : "00:00";
        return $"{date} {time}:00";
    }

    public string FormatDateOnlyForDisplay(string? dateStr, Func<string> todayLabel,
        Func<string> tomorrowLabel, Func<string> yesterdayLabel)
    {
        var dateOnly = ExtractDateOnly(dateStr);
        if (string.IsNullOrEmpty(dateOnly))
        {
            return string.Empty;
        }

        if (DateTime.TryParseExact(dateOnly, "yyyy-MM-dd", CultureInfo.InvariantCulture,
            DateTimeStyles.None, out var date))
        {
            var today = DateTime.Today;
            if (date.Date == today)
            {
                return todayLabel();
            }

            if (date.Date == today.AddDays(1))
            {
                return tomorrowLabel();
            }

            if (date.Date == today.AddDays(-1))
            {
                return yesterdayLabel();
            }
        }

        return dateOnly!;
    }

    public string FormatDateTimeForDisplay(string? dateTimeStr, Func<string> todayLabel,
        Func<string> tomorrowLabel, Func<string> yesterdayLabel)
    {
        if (string.IsNullOrEmpty(dateTimeStr))
        {
            return string.Empty;
        }

        var dateOnly = ExtractDateOnly(dateTimeStr);
        var timeOnly = ExtractTimeOnly(dateTimeStr);
        var dateLabel = FormatDateOnlyForDisplay(dateOnly, todayLabel, tomorrowLabel, yesterdayLabel);

        return string.IsNullOrEmpty(timeOnly)
            ? dateLabel
            : $"{dateLabel} {timeOnly}";
    }
}

File: apps/macos/Sources/Taskly/Services/DateParser.swift
import Foundation

/// Natural-language date/time parser. Faithful port of DateParser.cs.
///
/// Supported shorthands:
///   +10m (10 min), +2h, +1d, +1w, +1M (1 month = 30 days)
///   @now, @10am, @2pm, @10:30am, @22:30, @10am tomorrow/tmw, @8pm mon..sun
///   absolute: yyyy-MM-dd / yyyy/MM/dd / MM/dd/yyyy / dd/MM/yyyy
///   extractTimeCommand: pulls a trailing time command out of text
///     ("买牛奶 @10am" → ("买牛奶", "@10am"))
///
/// Parse results are 'yyyy-MM-dd HH:mm:ss'; date-only results 'yyyy-MM-dd'.
/// NOTE: unit case matters — m = minutes, M = month(30 days).
public struct DateParser: Sendable {
    public init() {}

    // Relative: +digits + unit (digits optional, default 1)
    private static let relativeRegex = try! NSRegularExpression(pattern: #"^(\d*)([mhdwM])$"#)
    // Time: @now / @10 / @10am / @10:30 / @10:30am / @22:30 / @22:30pm
    private static let timeRegex = try! NSRegularExpression(
        pattern: #"^(\d{1,2})(?::(\d{2}))?(am|pm)?$"#,
        options: [.caseInsensitive])
    // Trailing relative command: +10m / +2h …
    private static let extractRelativeRegex = try! NSRegularExpression(
        pattern: #"(?:^|\s)(\+\d+[mhdwM])(?:\s|$)"#)
    // Trailing @ command (with optional day-modifier suffix)
    private static let extractAtRegex = try! NSRegularExpression(
        pattern: #"@(?:now|\d{1,2}(?::\d{2})?(?:am|pm)?)(?:\s+(?:tomorrow|tmw|mon|tue|wed|thu|fri|sat|sun))?$"#,
        options: [.caseInsensitive])

    /// Parses input into 'yyyy-MM-dd HH:mm:ss' (or 'yyyy-MM-dd' for absolute
    /// dates). Returns nil when unparseable.
    public func parse(_ input: String) -> String? {
        let s = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !s.isEmpty else { return nil }

        if s.hasPrefix("+") { return parseRelativeDate(s) }
        if s.hasPrefix("@") { return parseAtTime(String(s.dropFirst())) }
        return parseAbsoluteDate(s)
    }

    /// Extracts a trailing time command from text. `timeCommand` is nil when
    /// nothing was extracted. Port of extractTimeCommand.
    public func extractTimeCommand(_ input: String) -> (text: String, timeCommand: String?) {
        guard !input.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return (input, nil)
        }

        var text = input
        var command: String?

        // Trailing @ command first
        if let m = Self.extractAtRegex.firstMatch(in: text, range: text.fullRange) {
            if let atIndex = text.firstIndex(of: "@") {
                command = String(text[atIndex...]).trimmingCharacters(in: .whitespaces)
                text = String(text[..<atIndex]).trimmingCharacters(in: CharacterSet(charactersIn: " \t\n\r"))
            }
            _ = m // match presence is what matters; bounds come from firstIndex(of:)
        }

        // Then trailing relative command
        if let m = Self.extractRelativeRegex.firstMatch(in: text, range: text.fullRange) {
            command = String(text[Range(m.range(at: 1), in: text)!])
            text = text.replacingCharacters(in: Range(m.range, in: text)!, with: "")
                .trimmingCharacters(in: .whitespaces)
        }

        return (text.trimmingCharacters(in: .whitespaces), command)
    }

    // ---------------- relative +N<unit> ----------------

    private func parseRelativeDate(_ input: String) -> String? {
        let body = String(input.dropFirst())
        guard let m = Self.relativeRegex.firstMatch(in: body, range: body.fullRange) else {
            return nil
        }

        let numStr = String(body[Range(m.range(at: 1), in: body)!])
        let amount = numStr.isEmpty ? 1 : (Int(numStr) ?? 0)
        let unit = String(body[Range(m.range(at: 2), in: body)!])

        let now = Date()
        let calendar = Calendar.current
        let result: Date?
        switch unit {
        case "m": result = calendar.date(byAdding: .minute, value: amount, to: now)
        case "h": result = calendar.date(byAdding: .hour, value: amount, to: now)
        case "d": result = calendar.date(byAdding: .day, value: amount, to: now)
        case "w": result = calendar.date(byAdding: .day, value: amount * 7, to: now)
        case "M": result = calendar.date(byAdding: .day, value: amount * 30, to: now)
        default: result = nil
        }

        return result.map { Self.formatFull($0) }
    }

    // ---------------- @time ----------------

    private func parseAtTime(_ input: String) -> String? {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)

        if trimmed.lowercased() == "now" {
            return Self.formatFull(Date())
        }

        // Split time body from optional day modifier (tomorrow/tmw/weekday)
        let parts = trimmed.split(separator: " ", maxSplits: 1, omittingEmptySubsequences: true)
        let timePart = parts.count > 0 ? String(parts[0]) : ""
        let modifier = parts.count > 1 ? String(parts[1]) : nil

        guard let m = Self.timeRegex.firstMatch(in: timePart, range: timePart.fullRange) else {
            return nil
        }

        var hour = Int(timePart[Range(m.range(at: 1), in: timePart)!]) ?? 0
        let min = m.range(at: 2).location != NSNotFound
            ? (Int(timePart[Range(m.range(at: 2), in: timePart)!]) ?? 0) : 0
        let ampm = m.range(at: 3).location != NSNotFound
            ? timePart[Range(m.range(at: 3), in: timePart)!].lowercased() : nil

        if ampm == "am" {
            if hour == 12 { hour = 0 }
        } else if ampm == "pm" {
            if hour < 12 { hour += 12 }
        }

        guard (0...23).contains(hour), (0...59).contains(min) else { return nil }

        let calendar = Calendar.current
        let now = Date()
        var date = calendar.startOfDay(for: now)
            .addingTimeInterval(TimeInterval(hour * 3600 + min * 60))

        if let modifier, !modifier.isEmpty {
            date = applyDayModifier(date, modifier)
        } else if date < now {
            // Past time with no modifier rolls over to tomorrow.
            date = calendar.date(byAdding: .day, value: 1, to: date) ?? date
        }

        return Self.formatFull(date)
    }

    private func applyDayModifier(_ date: Date, _ modifier: String) -> Date {
        let m = modifier.lowercased()
        if m == "tomorrow" || m == "tmw" {
            return Calendar.current.date(byAdding: .day, value: 1, to: date) ?? date
        }

        // Weekday: next occurrence (next week if today or already passed).
        let targetDay: Int?
        switch m {
        case "sun": targetDay = 1
        case "mon": targetDay = 2
        case "tue": targetDay = 3
        case "wed": targetDay = 4
        case "thu": targetDay = 5
        case "fri": targetDay = 6
        case "sat": targetDay = 7
        default: targetDay = nil
        }

        if let td = targetDay {
            let current = Calendar.current.component(.weekday, from: date)
            var diff = (td - current + 7) % 7
            if diff <= 0 { diff += 7 }
            return Calendar.current.date(byAdding: .day, value: diff, to: date) ?? date
        }

        return date
    }

    // ---------------- absolute dates ----------------

    private func parseAbsoluteDate(_ input: String) -> String? {
        let s = input.trimmingCharacters(in: .whitespacesAndNewlines)
        let formats = ["yyyy-MM-dd", "yyyy/MM/dd", "MM/dd/yyyy", "dd/MM/yyyy"]

        for fmt in formats {
            let formatter = DateFormatter()
            formatter.locale = Locale(identifier: "en_US_POSIX")
            formatter.timeZone = TimeZone.current
            formatter.dateFormat = fmt
            formatter.isLenient = false
            if let date = formatter.date(from: s) {
                let year = Calendar.current.component(.year, from: date)
                if year < ValidationHelper.minYear || year > ValidationHelper.maxYear {
                    return nil
                }
                return Self.string(from: date, format: "yyyy-MM-dd")
            }
        }

        return nil
    }

    // ---------------- display helpers ----------------

    /// Date part of 'yyyy-MM-dd HH:mm:ss' or 'yyyy-MM-dd' → 'yyyy-MM-dd'.
    public static func extractDateOnly(_ dateTimeStr: String?) -> String? {
        guard let dateTimeStr, !dateTimeStr.isEmpty else { return nil }
        return dateTimeStr.count >= 10 ? String(dateTimeStr.prefix(10)) : dateTimeStr
    }

    /// Time part 'HH:mm' of 'yyyy-MM-dd HH:mm:ss'. nil when absent.
    public static func extractTimeOnly(_ dateTimeStr: String?) -> String? {
        guard let dateTimeStr, !dateTimeStr.isEmpty, dateTimeStr.count >= 16 else { return nil }
        return String(dateTimeStr.dropFirst(11).prefix(5))
    }

    /// date + time → 'yyyy-MM-dd HH:mm:ss' (00:00:00 when no time).
    public static func combineDateTime(_ dateStr: String?, _ timeStr: String?) -> String {
        let date = extractDateOnly(dateStr) ?? Self.string(from: Date(), format: "yyyy-MM-dd")
        let time = (timeStr?.isEmpty == false) ? timeStr! : "00:00"
        return "\(date) \(time):00"
    }

    /// Date-only display (Today/Tomorrow/Yesterday/yyyy-MM-dd).
    public func formatDateOnlyForDisplay(
        _ dateStr: String?, todayLabel: String, tomorrowLabel: String, yesterdayLabel: String
    ) -> String {
        guard let dateOnly = Self.extractDateOnly(dateStr), !dateOnly.isEmpty else { return "" }

        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone.current
        formatter.dateFormat = "yyyy-MM-dd"
        if let date = formatter.date(from: dateOnly) {
            let calendar = Calendar.current
            let today = calendar.startOfDay(for: Date())
            let target = calendar.startOfDay(for: date)
            if target == today { return todayLabel }
            if target == calendar.date(byAdding: .day, value: 1, to: today) { return tomorrowLabel }
            if target == calendar.date(byAdding: .day, value: -1, to: today) { return yesterdayLabel }
        }

        return dateOnly
    }

    /// Date + time display ("Today 10:30").
    public func formatDateTimeForDisplay(
        _ dateTimeStr: String?, todayLabel: String, tomorrowLabel: String, yesterdayLabel: String
    ) -> String {
        guard let dateTimeStr, !dateTimeStr.isEmpty else { return "" }

        let dateOnly = Self.extractDateOnly(dateTimeStr)
        let timeOnly = Self.extractTimeOnly(dateTimeStr)
        let dateLabel = formatDateOnlyForDisplay(
            dateOnly, todayLabel: todayLabel, tomorrowLabel: tomorrowLabel,
            yesterdayLabel: yesterdayLabel)

        if let timeOnly, !timeOnly.isEmpty {
            return "\(dateLabel) \(timeOnly)"
        }
        return dateLabel
    }

    // ---------------- formatting utilities ----------------

    /// 'yyyy-MM-dd HH:mm:ss' in the local time zone (DateParser contract format).
    public static func formatFull(_ date: Date) -> String {
        string(from: date, format: "yyyy-MM-dd HH:mm:ss")
    }

    public static func string(from date: Date, format: String) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone.current
        formatter.dateFormat = format
        return formatter.string(from: date)
    }

    /// ISO-8601 with local offset, used for created_at (compatible with the
    /// .NET "o" round-trip format already in existing .db files).
    public static func createdAtNow() -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        formatter.timeZone = .current
        return formatter.string(from: Date())
    }

    /// "yyyy-MM-dd" → Date (today when nil/unparseable), for date pickers.
    public static func asDate(_ string: String?) -> Date {
        if let string {
            let formatter = DateFormatter()
            formatter.locale = Locale(identifier: "en_US_POSIX")
            formatter.timeZone = TimeZone.current
            formatter.dateFormat = "yyyy-MM-dd"
            if let date = formatter.date(from: string) {
                return date
            }
        }
        return Date()
    }

    /// "HH:mm" → Date (09:00 when nil/unparseable), for time pickers.
    public static func asTime(_ string: String?) -> Date {
        let calendar = Calendar.current
        let now = Date()
        if let string {
            let parts = string.split(separator: ":")
            if parts.count == 2, let hour = Int(parts[0]), let minute = Int(parts[1]),
               (0...23).contains(hour), (0...59).contains(minute) {
                return calendar.date(
                    bySettingHour: hour, minute: minute, second: 0, of: now) ?? now
            }
        }
        return calendar.date(bySettingHour: 9, minute: 0, second: 0, of: now) ?? now
    }
}

private extension String {
    var fullRange: NSRange { NSRange(startIndex..., in: self) }
}
